src/instantiate.jl

```diff
@@ -0,0 +1,116 @@
+"""
+    struct OptimizerWithAttributes
+        optimizer
+        params::Vector{Pair{AbstractOptimizerAttribute,<:Any}}
+    end
+
+Object grouping an optimizer constructor and a list of optimizer attributes.
+Instances are created with [`instantiate`](@ref).
+"""
+struct OptimizerWithAttributes
+    # Function that takes zero arguments and returns a new optimizer.
+    # The type of the function could be
+    # * `Function`: a function, or
+    # * `DataType`: a type, or
+    # * `UnionAll`: a type with missing parameters.
+    optimizer_constructor
+    params::Vector{Pair{AbstractOptimizerAttribute,Any}}
+end
+
+_to_param(param::Pair{<:AbstractOptimizerAttribute}) = param
+_to_param(param::Pair{String}) = RawParameter(param.first) => param.second
+function _to_param(param::Pair)
+    error("Expected an optimizer attribute or a string, got `$(param.first)` which is a `$(typeof(param.first))`.")
+end
+
+"""
+    OptimizerWithAttributes(optimizer_constructor, params::Pair...)
+
+Create an [`OptimizerWithAttributes`](@ref) with the parameters `params`.
+"""
+function OptimizerWithAttributes(optimizer_constructor, args::Vararg{Pair, N}) where N
+    if !applicable(optimizer_constructor)
+        error(_INSTANTIATE_NOT_CALLABLE_MESSAGE)
+    end
+    params = Pair{AbstractOptimizerAttribute,Any}[_to_param(arg) for arg in args]
+    return OptimizerWithAttributes(optimizer_constructor, params)
+end
+
+const _INSTANTIATE_NOT_CALLABLE_MESSAGE =
+    "The provided `optimizer_constructor` is invalid. It must be callable with zero" *
+    "arguments. For example, \"Ipopt.Optimizer\" or" *
+    "\"() -> ECOS.Optimizer()\". It should not be an instantiated optimizer " *
+    "like \"Ipopt.Optimizer()\" or \"ECOS.Optimizer()\"." *
+    "(Note the difference in parentheses!)"
+
+"""
+    _instantiate_and_check(optimizer_constructor)
+
+Create an instance of optimizer by calling `optimizer_constructor`.
+Then check that the type returned is an empty [`AbstractOptimizer`](@ref).
+"""
+function _instantiate_and_check(optimizer_constructor)
+    if !applicable(optimizer_constructor)
+        error(_INSTANTIATE_NOT_CALLABLE_MESSAGE)
+    end
+    optimizer = optimizer_constructor()
+    if !isa(optimizer, AbstractOptimizer)
+        error("The provided `optimizer_constructor` returned an object of type " *
+              "$(typeof(optimizer)). Expected a " *
+              "MathOptInterface.AbstractOptimizer.")
```

Comment:
Does this mean that the user can't provide a `MOI.ModelLike`? What's the reasoning for this? They might want to write out the model to file without solving.

Reply:
I copy pasted this from JuMP, see https://github.com/JuliaOpt/JuMP.jl/pull/2141. I agree that we can accept any `ModelLike`

Reply:
@mlubin what's your take on this as you added it in https://github.com/JuliaOpt/JuMP.jl/pull/2090

Reply:
If it's not an `AbstractOptimizer`, then it can't take `AbstractOptimizerAttribute`s, so it doesn't make sense to put it inside an `OptimizerWithAttributes`. It also doesn't make sense to pass something that doesn't return an optimizer to `JuMP.set_optimizer`, unless we want to rename the function.

For the use case of writing a model to a file without solving, you could either use `JuMP.direct_model` or build a model and copy it using `MOI.copy_to`. Do we need to support setting it via `set_optimizer`?

Reply:
@mlubin makes good points. I remove my objection.